geometry/meshop.cpp

```diff
@@ -252,6 +252,118 @@ void saveAsPly( const Mesh &mesh, const boost::filesystem::path &filepath){
     }
 }
 
+void saveAsPlyWithFeatures(const boost::filesystem::path& filepath,
+                           const Mesh& mesh,
+                           const math::Points3i& vertexColors,
+                           const math::Points3i& faceColors,
+                           const math::Points3& vertexNormals,
+                           const std::vector<int>& faceLabels)
+{
+    LOG(info2) << "Saving mesh to file <" << filepath << ">.";
+
+    std::ofstream out(filepath.string().c_str());
+    out.setf(std::ios::scientific, std::ios::floatfield);
+
+    unsigned validFaces(0);
+    for (const auto& face : mesh.faces)
+    {
+        if (!face.degenerate() && mesh.good(face)) validFaces++;
+    }
+
+    // create header
+    out << "ply\n"
+        << "format ascii 1.0\n"
+        << "comment generated by window-mesh\n"
+        << "element vertex " << mesh.vertices.size() << '\n'
+        << "property float x\n"
+        << "property float y\n"
+        << "property float z\n";
+
+    if (vertexColors.size())
+    {
+        utility::expect(vertexColors.size() == mesh.vertices.size(),
+                        "Length of vertexColors must be the same as number of "
+                        "mesh vertices.");
+        out << "property uchar red\n"
+            << "property uchar green\n"
+            << "property uchar blue\n";
+    }
+
+    if (vertexNormals.size())
+    {
+        utility::expect(vertexNormals.size() == mesh.vertices.size(),
+                        "Length of vertexNormals must be the same as number of "
+                        "mesh vertices.");
+        out << "property float nx\n"
+            << "property float ny\n"
+            << "property float nz\n";
+    }
+
+    out << "element face " << validFaces << '\n'
+        << "property list uchar int vertex_indices\n";
+
+    if (faceColors.size())
+    {
+        utility::expect(faceColors.size() == validFaces,
```

Comment:
I would prefer LOGTHROW as it logs the location of the error and it's stored in the log file as well. `utility::expect` makes it harder to track the error.